### PR TITLE
Cleanup duplicate barcodes

### PR DESCRIPTION
**Pull request by seichhorn**
Adding parameter to barcode filtering methods that removes barcodes that are putative duplicates in adjacent z planes, leaving just one instance of any duplicated barcodes.

### CODE REVIEW

File: docs/tasks.rst [hunk 1]
```diff
@@ -87,6 +91,9 @@ Description: Use an adaptive barcode to enrich the decode barcodes for the corre
 Parameters:
 
 * misidentification_rate -- The target misidentification rate, calculated as the number of blank barcodes per blank barcode divided by the number of coding barcodes per coding barcode.
+* remove\_zplane\_duplicates -- Remove putative duplicate barcode counts from adjacent z planes.
+* z\_planes\_above\_below -- If removing putative duplicate barcodes, number of adjacent z planes to consider, generally anything within 2 µm would be worth considering.
+* max\_centroid\_separation -- If removing putative duplicate barcodes, maximum euclidean distance in xy pixels that can separate the centroids of putative duplicates.
```
Comment (emanuega) [inline docs/tasks.rst]: Can we rename these parameters so it is clearer that they relate to the removal of duplicates across the z planes. Maybe "remove_z_duplicated_barcodes" "z_duplicate_axial_threshold" and "z_duplicate_lateral_threshold"

File: merlin/util/duplicatebcremoval.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+from scipy.spatial import cKDTree
+import networkx as nx
+import pandas as pd
```
Comment (emanuega) [inline merlin/util/duplicatebcremoval.py]: Can we rename this module so that in the future we could move some of the adaptive thresholding code into this file?

File: merlin/util/duplicatebcremoval.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from random import choice
+
+
+def cleanup_across_z(barcodes: pd.DataFrame, zPlanes: int, maxDist: float) \
+        -> pd.DataFrame:
+    """ Depending on the separation between z planes, spots from a single
+        molecule may be observed in more than one z plane. These putative
+        duplicates are removed based on supplied distance and z plane
```
Comment (emanuega) [inline merlin/util/duplicatebcremoval.py]: It could be useful to have a unit test for this function to make sure it behaves correctly in the edge cases and that we don't break it in the future.

File: merlin/util/duplicatebcremoval.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                      (pos >= i - zPlanes) & (pos <= i + zPlanes) & ~(pos == i)]
+        treeBC = barcodes[barcodes['z'] == z]
+        tree = cKDTree(treeBC.loc[:, ['x', 'y']].values)
+        for compZ in zToCompare:
+            queryBC = barcodes[barcodes['z'] == compZ]
+            dist, idx = tree.query(queryBC.loc[:, ['x', 'y']].values, k=1,
+                                   distance_upper_bound=maxDist)
+            currentHits = treeBC.index.values[idx[np.isfinite(dist)]]
+            comparisonHits = queryBC.index.values[np.isfinite(dist)]
+            graph.add_edges_from(list(zip(currentHits, comparisonHits)))
```
Comment (emanuega) [inline merlin/util/duplicatebcremoval.py]: I would expect this to only remove the barcodes assigned to the same barcode_id, but I don't find anywhere where barcode_id is being checked. Is this the correct assumption?

File: merlin/util/duplicatebcremoval.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    connectedComponents = [list(x) for x in
+                           list(nx.connected_components(graph))]
+    keptBarcodes = barcodes.loc[sorted([x[0] if len(x) == 1 else choice(x) for x
```
Comment (emanuega) [inline merlin/util/duplicatebcremoval.py]: I would prefer to either take the barcode that is brighter or to create a new barcode that is placed at the mean of the x,y, and z positions of the duplicates so that the decoding can be as deterministic as possible.

File: merlin/analysis/filterbarcodes.py [hunk 6]
```diff
@@ -316,7 +338,7 @@ def extreme_values(inputData: pandas.Series):
                         codingCounts, 'coding_counts', self)
 
 
-class AdaptiveFilterBarcodes(decode.BarcodeSavingParallelAnalysisTask):
+class AdaptiveFilterBarcodes(FilterBarcodes):
```
Comment (emanuega) [inline merlin/analysis/filterbarcodes.py]: This creates extra 'area_threshold', 'intensity_threshold', and 'distance_threshold' parameters in the parameter set for AdaptiveFilterBarcode that aren't relevant. I would prefer creating another parent class called AbstractFilterBarcodes that defines the shared parameters and can be used as the base class for both FilterBarcodes and AdaptiveFilterBarcodes.  The  _remove_zplane_duplicates function can all be moved to the utility module since I don't think the cleanup_across_z function is especially useful without this function.

File: merlin/util/duplicatebcremoval.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+from scipy.spatial import cKDTree
+import networkx as nx
+import pandas as pd
+from random import choice
+
+
+def cleanup_across_z(barcodes: pd.DataFrame, zPlanes: int, maxDist: float) \
+        -> pd.DataFrame:
```
Comment (emanuega) [inline merlin/util/duplicatebcremoval.py]: Can we assert that all the barcodes have the same barcode_id and rename the function to remove_zplane_duplicates_single_barcodeid to make sure this function isn't accidentally misused?

File: merlin/util/duplicatebcremoval.py (hunk 4 above)
Comment (seichhorn) [reply merlin/util/duplicatebcremoval.py]: In the implementation in filter barcodes the barcode database is loaded and then grouped by barcode id, the groups are run through util.duplicatebcremoval.cleanup_across_z individually so it would only remove barcodes assigned to the same id. I didn't put any logic explicitly enforcing this within the cleanup_across_z function, but I could put a check or warning or something if the user decided to run it on their own and didn't preprocess the data correctly?

Comment (emanuega) [reply merlin/util/duplicatebcremoval.py]: Sorry, I meant to delete this comment when I added the renaming suggesting back in.

**Comment by seichhorn** [comment]
I've addressed your comments. I also changed the .item() call that was causing a future warning from pandas, and the sumsignal task to default to false for apply high pass - there are some issues of erroneous background generated from this at fov borders when the signal is low that I dont fully understand.

File: docs/tasks.rst (hunk 1 above)
Comment (emanuega) [reply docs/tasks.rst]: Looks like these parameters were renamed in the code and just have to be updated in the documentation.

Comment (seichhorn) [reply docs/tasks.rst]: Yeah forgot them there. Should be good now.

**Comment by seichhorn** [comment]
Can you also merge 1.5 into master once this is in?
